### PR TITLE
fix(crosschain): protect gas payment swaps

### LINKED ISSUE #4565: fix: add slippage protection to gas payment Uniswap swaps
## Context

HackenProof report: https://dashboard.hackenproof.com/manager/companies/zetachain/zetachain-blockchain-software/reports/ZCNode-213

## Problem

All Uniswap V2 swap functions used for cross-chain gas payments hardcode `amountOutMin = BigIntZero` (`big.NewInt(0)`, defined at `x/fungible/keeper/evm.go:44`), providing zero slippage protection. The deadline is `big.NewInt(1e17)` (~3 billion years).

This enables a sandwich attack by any block-proposing validator: front-run the pool to skew the price, let the victim's gas payment swap execute at a worse rate, then back-run to extract profit. The victim overpays for gas and receives less on their cross-chain transfer.

**Two affected paths:**

**ZETA path** (`x/crosschain/keeper/gas_payment.go:388-444`):
- `QueryUniswapV2RouterGetZetaAmountsIn` reads spot reserves (line 388) — no TWAP
- `CallUniswapV2RouterSwapExactETHForToken` swaps with `amountOutMin=0` (line 421 → `system_contract.go:500`)
- Burns `amounts[1]` without checking against `outTxGasFee` (line 444, FIXME referencing #1303)

**ERC20 path** (`x/crosschain/keeper/gas_payment.go:219-294`):
- `QueryUniswapV2RouterGetZRC4ToZRC4AmountsIn` reads spot reserves (line 219)
- `CallUniswapV2RouterSwapExactTokensForTokens` swaps with `amountOutMin=0` (line 272 → `system_contract.go:382`)
- Output check at line 294 validates `gasObtained >= outTxGasFee` but cannot detect input-side inflation — the user overpays in input tokens while the check passes

## Fix

- Set `amountOutMin` to a percentage of the expected output (e.g. 95% of `outTxGasFee`) instead of zero in both swap functions
- ZETA path: add the missing output verification before burning (the FIXME at line 442, issue #1303)
- Set a reasonable deadline instead of `1e17` seconds

## Severity

Medium — sandwich attack by block-proposing validator, user overpays gas fees.

### PR DESCRIPTION
**Pull request by alan747271363-art**
## Summary
- add amountOutMin slippage protection for gas payment Uniswap swaps
- replace the effectively unbounded router deadline with a 30 minute deadline
- verify ZETA gas swaps receive enough gas ZRC20 before burning

## Tests
- go test -tags pebbledb,ledger,test ./x/fungible/keeper ./x/crosschain/keeper

Closes #4565

<h3>Greptile Summary</h3>

This PR hardens gas-payment Uniswap swaps on ZetaChain by adding slippage protection (`amountOutMin` at 5% below expected output), replacing the effectively infinite `1e17` deadline with a 30-minute block-time-relative deadline, and adding a new post-swap verification for the ZETA gas path to confirm at least `outTxGasFee` gas ZRC20 was received before burning.

- **Slippage guard**: `gasPaymentAmountOutMin` (95% of expected) is passed to both `swapExactTokensForTokens` and `swapExactETHForTokens` as `amountOutMin`; the ERC20 path also gains this guard (the post-swap check was already present).
- **ZETA path new check**: A strict `amounts[1] < outTxGasFee` guard is added before `CallZRC20Burn` in `PayGasInZetaAndUpdateCctx`, consistent with the existing ERC20 path behaviour.
- **Deadline fix**: `uniswapV2Deadline` replaces the 3-billion-year hardcoded value with a proper `blockTime + 30 min` computation across both swap helpers.

<h3>Confidence Score: 4/5</h3>

The core changes are correct and close a real vulnerability; the new protections are consistent across both payment paths and covered by new integration tests.

The `uniswapV2Deadline` fallback lands on Unix timestamp 1800 (January 1970) when block time is zero, silently breaking all swaps in that edge case. The `amountOutMin` at 95% and the post-swap check at 100% have mismatched thresholds that are undocumented, which could mislead future maintainers. Neither concern affects the production happy path.

x/fungible/keeper/system_contract.go (deadline fallback) and x/crosschain/keeper/gas_payment.go (dual-threshold design)

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| x/crosschain/keeper/gas_payment.go | Adds slippage protection (5% amountOutMin) for ERC20 and ZETA gas swaps, and adds a new strict post-swap check for the ZETA path; the dual-threshold design (95% EVM guard + 100% Cosmos check) is intentional but undocumented. |
| x/fungible/keeper/system_contract.go | Replaces hardcoded `1e17` deadline and zero amountOutMin with a proper 30-minute block-time-relative deadline and caller-supplied amountOutMin; the negative-deadline fallback resolves to Unix timestamp 1800, which would fail all swaps in zero-time test contexts. |
| x/fungible/keeper/system_contract_test.go | Test signatures updated for the new amountOutMin parameter; new sub-cases added to verify that exceeding amountOutMin causes ErrContractCall for both swap variants. |
| x/crosschain/types/expected_keepers.go | Interface signatures for both Uniswap swap helpers updated to include the new amountOutMin parameter. |
| testutil/keeper/mocks/crosschain/fungible.go | Auto-generated mock correctly regenerated to reflect the new amountOutMin parameter in both swap functions. |
| testutil/keeper/crosschain.go | Mock setup for PayGasAndUpdateCCTX updated to add an extra mock.Anything arg matching the new amountOutMin parameter. |

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
x/fungible/keeper/system_contract.go:22-28
**Fallback deadline produces an always-expired timestamp**

When `ctx.BlockTime()` is zero (the Go zero-value `time.Time{}`), `ctx.BlockTime().Add(30 * time.Minute).Unix()` is a large negative number, triggering the fallback. The fallback sets `deadline = int64(uniswapV2SwapDeadline.Seconds())` which is `1800` — Unix timestamp January 1, 1970, 00:30 UTC. Any Uniswap call carrying this value will immediately revert with `UniswapV2Router: EXPIRED`, since `block.timestamp > 1800` in every real or non-trivial test context. The intent of the fallback appears to be "use a safe default when the block time is unavailable," but the value chosen is the opposite of safe.

### Issue 2 of 2
x/crosschain/keeper/gas_payment.go:38-45
**`amountOutMin` at 95% is superseded by the 100% post-swap checks**

`gasPaymentAmountOutMin` allows up to 5% slippage at the EVM level, but both gas payment paths immediately follow the swap with a strict check (`amounts[x].Cmp(outTxGasFee.BigInt()) == -1`) that requires 100% of the expected output. Any swap output in the 95–100% range passes the EVM guard but is rejected by the Cosmos-level check, making the 5% tolerance effectively unreachable. A comment explaining this dual-layer design (EVM guard stops catastrophic slippage; Cosmos check enforces accounting exactness and reverts cleanly) would help future readers understand why the thresholds are intentionally mismatched, and prevent someone from "fixing" the `amountOutMin` to 100% while silently losing the EVM-level protection.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(crosschain): protect gas payment swa..."](https://github.com/zeta-chain/node/commit/f2377e9ceda87e4b1ea0c81ad6b0a0b9046e7c91) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31997697)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: x/fungible/keeper/system_contract.go [hunk 1]
```diff
@@ -16,6 +17,16 @@ import (
...
 
+const uniswapV2SwapDeadline = 30 * time.Minute
+
+func uniswapV2Deadline(ctx sdk.Context) *big.Int {
+	deadline := ctx.BlockTime().Add(uniswapV2SwapDeadline).Unix()
+	if deadline < 0 {
+		deadline = int64(uniswapV2SwapDeadline.Seconds())
+	}
+	return big.NewInt(deadline)
+}
```
[the remark this answers is not in the document]
Comment (alan747271363-art) [reply x/fungible/keeper/system_contract.go:35]: Addressed on the current branch: `uniswapV2Deadline` now explicitly handles zero block time and negative deadlines with a far-future fallback deadline (`100_000_000_000_000_000`) instead of `1800`, so the fallback is not immediately expired.

File: x/crosschain/keeper/gas_payment.go [hunk 2]
```diff
@@ -30,6 +30,20 @@ type ChainGasParams struct {
...
+)
+
+func gasPaymentAmountOutMin(amountOut *big.Int) *big.Int {
+	if amountOut == nil || amountOut.Sign() <= 0 {
+		return big.NewInt(0)
+	}
+
+	amountOutMin := new(big.Int).Mul(amountOut, big.NewInt(gasPaymentBpsDenominator-gasPaymentSwapSlippageBps))
+	return amountOutMin.Div(amountOutMin, big.NewInt(gasPaymentBpsDenominator))
+}
```
[the remark this answers is not in the document]
Comment (alan747271363-art) [reply x/crosschain/keeper/gas_payment.go:47]: Addressed on the current branch with a comment above the 95% minimum: the EVM-level `amountOutMin` is kept as a catastrophic-slippage guard, while the later full-fee check preserves exact CCTX accounting.
